Make hardtanh as CompositeImplicitAutograd

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67475

Summary: In the process of making hardtanh a structured kernel, we
discovered that hardtanh is really just clamp with a default min and max
val, so it makes more sense to make hardtanh as
CompositeImplicitAutograd. Because clamp_out is missing a quantized
version, we also need to add one in this PR.

Test Plan:
```
pytest test/test_ops.py -k "hardtanh"
pytest test/test_ops.py -k "clamp"
```

desertfire committed:
Make hardtanh into structured kernel

Test:
`pytest test/test_ops.py -k "hardtanh"`

[ghstack-poisoned]

desertfire committed:
Update on "Make hardtanh into structured kernel"


Test:
`pytest test/test_ops.py -k "hardtanh"`

[ghstack-poisoned]

desertfire commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -8616,33 +8616,14 @@
   python_module: nn
-  dispatch:
-    CPU, CUDA: hardtanh_out
-    QuantizedCPU: hardtanh_out_quantized_cpu
```
@bdhirsh do you think quantized would be a problem here? I see `clamp` has a quantized method, but not `clamp_out`, https://github.com/pytorch/pytorch/blob/9cdd1d7e489b17ba86aafdadef37eecbf0a64917/aten/src/ATen/native/native_functions.yaml#L1064

bdhirsh replied:
Oh, you're right. So I think your two options are:

(a) Add a `QuantizedCPU` kernel for `clamp.out`. This should hopefully be pretty easy, since you can just use pretty much the same kernel currently used for `hardtanh.out`. Then you can fully remove the `hardtanh.out` entry

(b) Otherwise just leave the QuantizedCPU entry for `hardtanh.out`

bdhirsh commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -62,6 +62,7 @@
     ("aten::grid_sampler_2d_backward", datetime.date(2021, 10, 21)),
     ("prim::TensorExprDynamicGuard", datetime.date(2021, 11, 20)),
     ("caffe2::", datetime.date(2021, 10, 23)),
+    ("aten::hardtanh", datetime.date(2021, 11, 1)),
```
btw the risk of adding an optimistic date here is that if this merges to master within 24 hours after the date specified, the BC test will start failing on master (for a day, before it stops).

I usually pessimistically put something ~2-3 weeks in the future, maybe December 1st

desertfire committed:
Update on "Make hardtanh as CompositeImplicitAutograd"


Summary: In the process of making hardtanh a structured kernel, we
discovered that hardtanh is really just clamp with a default min and max
val, so it makes more sense to make hardtanh as
CompositeImplicitAutograd. Because clamp_out is missing a quantized
version, we also need to add one in this PR.

Test Plan:
```
pytest test/test_ops.py -k "hardtanh"
pytest test/test_ops.py -k "clamp"
```

[ghstack-poisoned]

desertfire committed:
Update on "Make hardtanh as CompositeImplicitAutograd"


Summary: In the process of making hardtanh a structured kernel, we
discovered that hardtanh is really just clamp with a default min and max
val, so it makes more sense to make hardtanh as
CompositeImplicitAutograd. Because clamp_out is missing a quantized
version, we also need to add one in this PR.

Test Plan:
```
pytest test/test_ops.py -k "hardtanh"
pytest test/test_ops.py -k "clamp"
```

[ghstack-poisoned]

desertfire committed:
Update on "Make hardtanh as CompositeImplicitAutograd"


Summary: In the process of making hardtanh a structured kernel, we
discovered that hardtanh is really just clamp with a default min and max
val, so it makes more sense to make hardtanh as
CompositeImplicitAutograd. Because clamp_out is missing a quantized
version, we also need to add one in this PR.

Test Plan:
```
pytest test/test_ops.py -k "hardtanh"
pytest test/test_ops.py -k "clamp"
```

[ghstack-poisoned]

desertfire commented:
@bdhirsh How should we handle the sync with pytorch-xla? (see the CI failure on  pytorch_xla_linux_bionic_py3_6_clang9_build)

desertfire commented on aten/src/ATen/native/quantized/cpu/qclamp.cpp:
```diff
@@ -128,30 +128,21 @@ Tensor clamp_quantized_cpu(
   Tensor qy;
-  qy = quantized_clamp_impl(self, min, max);
+  qy = clamp_quantized_cpu(self, min, max);
   // This can be optimized in a future PR if it becomes a bottleneck.
```
This comment makes me nervous. I also found comment like https://github.com/pytorch/pytorch/blob/04fe4382ec8e6031d353530d6b8446ac9c9f507e/torch/csrc/jit/passes/quantization/helper.cpp#L169. I hope I am not opening a can of worms here, LOL

bdhirsh replied:
That commented also added 1.5 years ago 😬

Yeah, it looks like the current state is that out of these 4 ops: `hardtanh`, `hardtanh_`, `clamp`, `clamp_`, quantized currently supports all but the last one. And technically, this PR will add support for `clamp_`.

I think... we can count on CI to alert us if there's a really good reason for not adding support for that 4th variant?

bdhirsh commented:
Yep, this PR needs xla-side changes. @JackCaoG when you get a chance do you mind making the xla change? This PR should let us completely remove the `hardtanh` and `hardtanh_backward` ops in xla, now that we're making hardtanh composite.

JackCaoG commented:
@bdhirsh Sounds good, will take a look.

desertfire committed:
Update on "Make hardtanh as CompositeImplicitAutograd"


Summary: In the process of making hardtanh a structured kernel, we
discovered that hardtanh is really just clamp with a default min and max
val, so it makes more sense to make hardtanh as
CompositeImplicitAutograd. Because clamp_out is missing a quantized
version, we also need to add one in this PR.

Test Plan:
```
pytest test/test_ops.py -k "hardtanh"
pytest test/test_ops.py -k "clamp"
```

[ghstack-poisoned]

wconstab commented:
@desertfire did we confirm that nvfuser folks are actually ok with us not capturing hardtanh explicitly, and instead capturing a min/max op? If so, no objection from me.

desertfire committed:
Update on "Make hardtanh as CompositeImplicitAutograd"


Summary: In the process of making hardtanh a structured kernel, we
discovered that hardtanh is really just clamp with a default min and max
val, so it makes more sense to make hardtanh as
CompositeImplicitAutograd. Because clamp_out is missing a quantized
version, we also need to add one in this PR.

Test Plan:
```
pytest test/test_ops.py -k "hardtanh"
pytest test/test_ops.py -k "clamp"
```

[ghstack-poisoned]

desertfire committed:
Update on "Make hardtanh as CompositeImplicitAutograd"


Summary: In the process of making hardtanh a structured kernel, we
discovered that hardtanh is really just clamp with a default min and max
val, so it makes more sense to make hardtanh as
CompositeImplicitAutograd. Because clamp_out is missing a quantized
version, we also need to add one in this PR.

Test Plan:
```
pytest test/test_ops.py -k "hardtanh"
pytest test/test_ops.py -k "clamp"
```

[ghstack-poisoned]

wconstab commented:
> @desertfire did we confirm that nvfuser folks are actually ok with us not capturing hardtanh explicitly, and instead capturing a min/max op? If so, no objection from me.

@desertfire just bumping this comment

bdhirsh commented:
> did we confirm that nvfuser folks are actually ok with us not capturing hardtanh explicitly, and instead capturing a min/max op? If so, no objection from me.

I think it would still be generally useful to mark `hardtanh` as composite, since that allows other extensions (like functorch, other backends, etc) to not have to worry about it. If a specific backend like nvfuser wants to handle a composite op explicitly, I think we should be able to get around that by registering a kernel to the `AutogradLazy` key directly?

wconstab commented:
ok, that's true.  SGTM then.

desertfire commented:
> @bdhirsh Sounds good, will take a look.

@JackCaoG, should we land your https://github.com/pytorch/xla/pull/3205/ first?

JackCaoG commented:
@desertfire Let me rebase my old pr and rerun the test. I usually let pytorch pr to be merged first and I merge pytorch/xla pr right after.

desertfire commented:
But that means our CI would fail for a while, right? I think you xla PR can go in first because the fallback should kick in after you remove hardtanh_backward from pytorch/xla.

JackCaoG commented:
@desertfire pytorch/xla has test to check whether a op will be execute on XLA device or not. If I just merge my pr without yours, that check will fail due to fallback. One thing you could do is to ping me before merging your pr, I usually try to merge corresponding pt/xla pr right after pytorch one and it usually don't cause too much dtsiriburance.

ngimel commented on aten/src/ATen/native/cuda/Activation.cu:
```diff
@@ -322,16 +322,6 @@ void shrink_backward_kernel(TensorIteratorBase& iter, const Scalar& value) {
-  AT_DISPATCH_FLOATING_TYPES_AND(at::ScalarType::Half, iter.dtype(), "hardtanh_backward_cuda", [&]() {
-    auto min_val = min.to<scalar_t>();
-    auto max_val = max.to<scalar_t>();
-    gpu_kernel(iter, [min_val, max_val]GPU_LAMBDA(scalar_t a, scalar_t b) -> scalar_t {
```
If you remove this and instead rely on clamp_backward, the perf will regress significantly (clamp_backward formula is 
```
    auto zero = at::scalar_tensor(0., grad.options());
    return where((self >= *min).logical_and_(self <= *max), grad, zero);
```
 and it requires multiple passes over data, whereas this kernels traverses the data just once.

desertfire replied:
I see. Then is the solution to add a specialized CUDA version of clamp_backward? I don't see clamp_backward has an entry in native_functions.yaml, so I am not sure how to REGISTER_DISPATCH for it.

ngimel replied:
Yeah, probably specialized cuda and cpu versions of clamp_backward are needed (similar to hardtanh_backward that we have). clamp_backward is in FunctionsManual.cpp, and we need to move it to native_functions from there, otherwise we can't have it as a structured kernel? @albanD do we have any policy on which backward functions live exclusively in FunctionsManual, and which are in native_functions? Currently it looks like a mix of both, and probably makes sense to move more things to native_functions? Also we'd need an entry for `clamp_backward` itself in the derivatives.yaml, similar to how we have it for hardtanh_backward.

albanD replied:
Ho my bad! I though clamp_backward had a custom kernel...
So it is not that much duplicate code then. We just want to route everything to this kernel (optionally renaming it clamp_backward).

There isn't a strict rule. But for CompositeImplicit functions, there is almost no advantage to be in the dispatcher but there is the big disadvantage to have to follow native op FC/BC rules. So we usually keep them in FunctionsManual.
For non-composite implicit stuff where you need either a custom double backward or kernels, we do an op to easily use the codegen.

bdhirsh replied:
From talking to @albanD offline, it sounds like we should just rename the `hardtanh_backward` native_function op to `clamp_backward` (and then kill the custom `clamp_backward` function in `FunctionsManual.cpp).

Logically, this PR would then be doing 3 things:
- Making `clamp`'s backward more efficient by using `hardtanh_backward` (since its backward is really just the same as hardtanh_backward, but the second is written more efficiently)
- Renaming `hardtanh_backward` -> `clamp_backward`
- Making `hardtanh` composite

ngimel replied:
`clamp_backward` also handles cases where `min` or `max` are not specified, so some care is needed for that.

desertfire replied:
Also `clamp` supports tensor as `min` or `max`, but `hardtanh` only support scalar `min` and `max`. At this point, I feel it's worth to split this PR into, 1. improve `clamp_backward`'s efficiency; 2. make `hardtanh` composite and delegate `hardtanh_backward` to `clamp_backward`.

ngimel commented:
`clamp_backward` also handles cases where `min` or `max` are not specified, so some care is needed for that.
